### DISCUSSION TITLE
How to run the demo on CPU?

### DISCUSSION THREAD

**Discussion by LiuCMU** [open-mmlab/mmediting#1483] [How-to] [answered]
Hi, I'm trying to run this [demo](https://mmediting.readthedocs.io/en/latest/demo.html#restoration-image).

But I got the following error message due to a lack of memory in GPU.
```
(py39) liucmu@storm:~/mmediting$ python demo/restoration_demo.py     configs/restorers/esrgan/esrgan_x4c64b23g32_g1_400k_div2k.py     esrgan_x4c64b23g32_1x16_400k_div2k_20200508-f8ccaf3b.pth    "/storage/users/jack/MS_ML_datasets/img_deblur/test/blur_gamma/GOPR0396_11_00_000001.png"  demo_out.png
2022-11-29 01:19:10,530 - mmedit - INFO - load checkpoint from torchvision path: torchvision://vgg19
load checkpoint from local path: esrgan_x4c64b23g32_1x16_400k_div2k_20200508-f8ccaf3b.pth
Traceback (most recent call last):
  File "/home/liucmu/mmediting/demo/restoration_demo.py", line 62, in <module>
    main()
  File "/home/liucmu/mmediting/demo/restoration_demo.py", line 53, in main
    output = restoration_inference(model, args.img_path)
  File "/home/liucmu/mmediting/mmedit/apis/restoration_inference.py", line 46, in restoration_inference
    result = model(test_mode=True, **data)
  File "/home/liucmu/miniconda3/envs/py39/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/liucmu/miniconda3/envs/py39/lib/python3.9/site-packages/mmcv/runner/fp16_utils.py", line 110, in new_func
    return old_func(*args, **kwargs)
  File "/home/liucmu/mmediting/mmedit/models/restorers/srgan.py", line 95, in forward
    return self.forward_test(lq, gt, **kwargs)
  File "/home/liucmu/mmediting/mmedit/models/restorers/basic_restorer.py", line 140, in forward_test
    output = self.generator(lq)
  File "/home/liucmu/miniconda3/envs/py39/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/liucmu/mmediting/mmedit/models/backbones/sr_backbones/rrdb_net.py", line 169, in forward
    self.conv_up2(F.interpolate(feat, scale_factor=2, mode='nearest')))
  File "/home/liucmu/miniconda3/envs/py39/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/liucmu/miniconda3/envs/py39/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 447, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/home/liucmu/miniconda3/envs/py39/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 443, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
RuntimeError: CUDA out of memory. Tried to allocate 3.52 GiB (GPU 0; 7.80 GiB total capacity; 4.82 GiB already allocated; 1.03 GiB free; 5.06 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```
Since my GPU memory is limited (8GB), and CPU memory is plenty (128G). Is it possible to run this demo on CPU so that I can get rid of the memory issues? If so, how could I do it? Thank you very much!

**Comment by LiuCMU** [comment]
oh, I figured it out: `device` can be modified in `demo/restoration_demo.py`.

**Comment by zengyh1900** [comment, marked as answer]
> python demo/restoration_demo.py

Hi @LiuCMU,
you can also try by running `CUDA_VISIBLE_DEVICES=-1 python demo/restoration_demo.py  xxxxx` so that it can not detect GPU device and will run it on CPU. 

btw, we have upgraded the demo experience in branch dev-1.x, which is our develop branch. 
In dev-1.x you can try the demo by API in notebook https://github.com/open-mmlab/mmediting/blob/dev-1.x/demo/mmediting_inference_tutorial.ipynb
or try it by CLI https://github.com/open-mmlab/mmediting/tree/dev-1.x/demo

Any feedback are welcome!

**Comment by LiuCMU** [comment]
Hey @zengyh1900 
Thank you very much for the timely response!

The documentation version is a little confusing for me. I followed the [documentation](https://mmediting.readthedocs.io/en/latest/install.html) to install MMEditing. It suggests the "latested" version in the bottom-left, which I thought is the 1.x version. But I ended up with `0.16.0` after installation. I guess it will be helpful for future users if the installation documentation could be updated.

A side note: I tried `ESRGAN` demo to do a deblur task, but the result didn't look promising (see below). This is probably due to the fact that deblur and super-resolution are two very different tasks (sorry I don't have much experience in CV). I will continue to search for other methods. 

Anyway, thanks for the great job. MMEdit is an excellent project.
<img width="1578" alt="image" src="https://user-images.githubusercontent.com/60156077/204427747-d18ecfd4-1429-422b-99a8-21ddb03bfbae.png">

**Reply by zengyh1900** [reply]
Hi @LiuCMU 

1. Thank you for your suggestion! Yes, you're right! The `latest` tag is used for 0.x version, and the documentation for 1.x version can be found in the `dev-1.x` tag. 
2. MMEditing 1.x is now in the open-beta phase. We will release it in 2022.12.31, at that time we will update our documentation tag. 
3. `ESRGAN` is trained on low-to-high resolution image pairs, which is quite different from deblur task. Therefore, the results are expected. I think this repo https://github.com/ljzycmd/SimDeblur can provide you some deblur models for your case. BTW, we are calling for the community to contribute some related deblur works to MMEditing. Please feel free to contact us if you're interested in it!

**Reply by LiuCMU** [reply]
Got it, thank you! @zengyh1900
